Rootfs automount reload race: umount2(<rootfs>/proc) in host ns vs new prototype's mount() (flaky test_go_isolation_rootfs_automount_tmpfs)

## Summary
Under a rapid same-rootfs reload loop, the previous worker generation's mount teardown races the new prototype's `proc` mount on the **same host path `<rootfs>/proc`**, with no serialization. The loser hits `mount(...proc...) ENOENT`, the prototype exits 1, a respawn storm follows, the generation never converges, and dangling app↔router fds trip the fd-leak teardown check.

This surfaces as a flaky CI failure in `test_go_isolation_rootfs_automount_tmpfs` / `..._regression` — see PR #78, run `27793786107` (`test (go-1.25)` red, `test (go-1.26)` green on the *identical* daemon and `ns_inspect` app). Original flaky context: #60.

Credit: root cause first pinned by @andypost in https://github.com/freeunitorg/freeunit/pull/78#pullrequestreview-4544602879.

## Symptom (failing run, go-1.25)
```
[alert] mount("none", "/tmp/unit-test-…/proc", "proc", 2097162, "") (2: No such file or directory)
[notice] process … exited with code 1
FAILED test_go_isolation_rootfs_automount_tmpfs - mounts did not converge after reload (status=None)
ERROR  ...                                      - AssertionError: descriptors leak router (assert 53 <= 0)
```

## The race
Each reload swaps the worker generation; two operations touch `<rootfs>/proc` concurrently:
- **new prototype (child):** `mkdir` + `mount("none", <rootfs>/proc, "proc", …)` — `nxt_isolation_prepare_rootfs()` (`src/nxt_application.c:562` → `src/nxt_isolation.c:778`)
- **old-gen teardown (main process, host mount ns):** `umount2(<rootfs>/proc, MNT_DETACH)` — `nxt_isolation_unmount_all()` (`src/nxt_isolation.c:744`, registered as `isolation.cleanup` at `:602`)

The lazy detach lands on the path the new generation is mid-mounting → ENOENT → prototype exit 1 → respawn storm → `status=None` + leaked port fds. The Go toolchain version only shifts *when* the new `mount()` fires relative to the old `umount2()`; 1.25 hits the window, 1.26 misses it. This is **timing, not a Go-version code path**.

## Why the existing mitigation is insufficient
The private-mount-namespace mitigation is already in place (`cceba1e9` *private mount namespace*, `2063c666` *depin reload race*; tests load with `namespaces: {mount: True}` and `skip_alert` the ENOENT; proc-before-tmpfs ordering in `nxt_isolation_mount_compare`, `src/nxt_isolation.c:716`). It is **necessary but not sufficient**, because of the ordering in the child setup:

1. `nxt_isolation_prepare_rootfs` mounts `<rootfs>/proc` — `src/nxt_application.c:562`
2. only **afterwards** `nxt_isolation_change_root` → `nxt_isolation_pivot_root` severs propagation with `mount("", "/", MS_SLAVE|MS_REC, "")` — `src/nxt_isolation.c:878`

`unshare(CLONE_NEWNS)` alone does not detach mount propagation. On a systemd host, `/` (and `/tmp`, where `temp_dir` lives) is `MS_SHARED` by default, so in the window between `unshare` and `pivot_root` the new generation's proc mount is still propagation-linked to the host — exactly while the host-side teardown `umount2`s the same path. `MS_SLAVE` arrives too late to close the window.

## Proposed fixes (pick one / combine)
1. **Sever propagation before mounting:** make the rootfs/`/` `MS_REC|MS_PRIVATE` (or `MS_SLAVE`) **before** `nxt_isolation_prepare_rootfs`, not inside `pivot_root` after it.
2. **Don't host-umount per-worker automounts:** when the worker has a private mount ns, let its automounts die with the namespace instead of `umount2`-ing them from the host ns in `nxt_isolation_unmount_all`.
3. **Serialize reload teardown:** ensure the old generation's mount teardown completes before the new prototype starts mounting the same rootfs.

## Test-side follow-up (PR #78)
- Until the C fix lands, mark `test_go_isolation_rootfs_automount_tmpfs[_regression]` `xfail`/skip on go-1.25, referencing this issue, to unblock the CI-workflow refactor (do **not** drop go-1.25 from the matrix — it is a supported runtime per `pkg/eol.json`).
- The `_regression` variant (×100 reload loop) is a CI time-bomb regardless of toolchain; consider gating it behind an opt-in marker.

## Environment
- ubuntu-latest GitHub runner, root (`is_su=True`), rootfs isolation + `namespaces: {mount: True}` + tmpfs automount toggled per reload.
- Reproduces deterministically on the go-1.25 toolchain; not observed on go-1.26.

--- a/CHANGES
+++ b/CHANGES
@@ -35,6 +35,13 @@ Changes with FreeUnit 1.35.6                                     25 Jun 2026
     *) Bugfix: remove unreachable Protocol::HttpJson dead-code arm from the
        OTel exporter; only HttpBinary was ever selected.
 
+    *) Bugfix: fix a rootfs automount reload race: sever the worker's mount
+       namespace propagation (MS_REC|MS_PRIVATE) before mounting and let a
+       per-worker automount be reaped with the namespace instead of being
+       unmounted from the host, so a configuration reload no longer races the
+       previous worker generation's teardown against the new prototype's
+       procfs mount (freeunitorg/freeunit#83).
+
 
 Changes with FreeUnit 1.35.5                                     29 May 2026
 

--- a/src/nxt_isolation.c
+++ b/src/nxt_isolation.c
@@ -757,6 +757,22 @@ nxt_isolation_unmount_all(nxt_task_t *task, nxt_process_t *process)
 
     nxt_debug(task, "unmount all (%s)", process->name);
 
+#if (NXT_HAVE_CLONE_NEWNS)
+    /*
+     * A worker that ran in its own mount namespace had all of its mounts
+     * reaped by the kernel when that namespace died on process exit.
+     * Unmounting the same host paths here is both pointless (the mounts are
+     * not present in the host namespace — this is the EINVAL we used to log)
+     * and unsafe: it races the next worker generation's prototype, which
+     * mounts the identical <rootfs>/proc path (freeunitorg/freeunit#83).
+     * The non-namespaced (chroot) path below still needs the host-side
+     * umount2, so only skip when CLONE_NEWNS was in effect.
+     */
+    if (nxt_is_clone_flag_set(process->isolation.clone.flags, NEWNS)) {
+        return;
+    }
+#endif
+
     automount = &process->isolation.automount;
     mounts = process->isolation.mounts;
     n = mounts->nelts;
@@ -790,6 +806,30 @@ nxt_isolation_prepare_rootfs(nxt_task_t *task, nxt_process_t *process)
 
     n = mounts->nelts;
     mnt = mounts->elts;
+
+#if (NXT_HAVE_CLONE_NEWNS)
+    /*
+     * unshare(CLONE_NEWNS) copies the host mount tree *and* its propagation
+     * type.  On a systemd host "/" (and "/tmp", where a rootfs often lives) is
+     * MS_SHARED, so the new namespace starts as a peer of the host and the
+     * mounts below would propagate out to it.  pivot_root() only severs
+     * propagation later (nxt_isolation_pivot_root), so during this loop a
+     * concurrent host-side umount2(<rootfs>/proc) from the previous worker
+     * generation's teardown races our mount of the same path, yielding a
+     * transient "mount(...proc...) ENOENT" and a respawn storm
+     * (freeunitorg/freeunit#83).
+     *
+     * Detach the whole namespace from the host peer group *before* mounting.
+     * MS_PRIVATE (not MS_SLAVE) is required: a slave still receives propagation
+     * from its former master, so a host umount would propagate back in.
+     */
+    if (nxt_is_clone_flag_set(process->isolation.clone.flags, NEWNS)) {
+        if (nxt_slow_path(mount("", "/", "", MS_REC | MS_PRIVATE, "") != 0)) {
+            nxt_alert(task, "mount(\"/\", MS_REC|MS_PRIVATE) %E", nxt_errno);
+            return NXT_ERROR;
+        }
+    }
+#endif
 
     for (i = 0; i < n; i++) {
         dst = mnt[i].dst;

